File: fe1-web/model/network/method/message/data/witness/WitnessMessage.ts
import { Hash, Signature } from 'model/objects';
import { ProtocolError } from 'model/network/ProtocolError';
import { ActionType, MessageData, ObjectType } from '../MessageData';

export class WitnessMessage implements MessageData {
  public readonly object: ObjectType = ObjectType.MESSAGE;

  public readonly action: ActionType = ActionType.WITNESS;

  public readonly message_id: Hash;

  public readonly signature: Signature;

  constructor(msg: Partial<WitnessMessage>) {
    if (!msg.message_id) throw new ProtocolError('Undefined \'message_id\' parameter encountered during \'WitnessMessage\'');
    this.message_id = msg.message_id;

    if (!msg.signature) throw new ProtocolError('Undefined \'signature\' parameter encountered during \'WitnessMessage\'');
    // FIXME verify signature without the public key
    // available? 0.o + uncomment 3 tests in "FromJson" test suite
    this.signature = msg.signature;
  }

  public static fromJson(obj: any): WitnessMessage {
    // FIXME add JsonSchema validation to all "fromJson"
    const correctness = true;

    return correctness
      ? new WitnessMessage({
        ...obj,
        message_id: new Hash(obj.message_id),
        signature: new Signature(obj.signature),
      })
      : (() => { throw new ProtocolError('add JsonSchema error message'); })();
  }
}

File: be2-scala/src/main/scala/ch/epfl/pop/model/network/method/message/data/witness/WitnessMessage.scala
package ch.epfl.pop.model.network.method.message.data.witness

import ch.epfl.pop.model.network.Parsable
import ch.epfl.pop.model.network.method.message.data.ActionType.ActionType
import ch.epfl.pop.model.network.method.message.data.ObjectType.ObjectType
import ch.epfl.pop.model.network.method.message.data.{ActionType, MessageData, ObjectType}
import ch.epfl.pop.model.objects.{Hash, Signature}

import ch.epfl.pop.jsonNew.MessageDataProtocol._
import spray.json._

case class WitnessMessage(
                           message_id: Hash,
                           signature: Signature,
                         ) extends MessageData {
  override val _object: ObjectType = ObjectType.MESSAGE
  override val action: ActionType = ActionType.WITNESS
}

object WitnessMessage extends Parsable {
  def apply(
             message_id: Hash,
             signature: Signature,
           ): WitnessMessage = {
    new WitnessMessage(message_id, signature)
  }

  override def buildFromJson(messageData: MessageData, payload: String): WitnessMessage =
    payload.parseJson.asJsObject.convertTo[WitnessMessage]
}
